File: java/src/main/java/io/github/rahulbsw/idgenkit/Ulid.java
package io.github.rahulbsw.idgenkit;

import java.util.Arrays;
import java.util.UUID;

/**
 * Universally Unique Lexicographically Sortable Identifier.
 *
 * <p>128 bits: a 48-bit millisecond Unix timestamp followed by 80 bits of randomness, encoded as
 * 26 characters of Crockford base32. See https://github.com/ulid/spec. Natural ordering (unsigned)
 * matches both creation time and string order.
 */
public final class Ulid implements Comparable<Ulid> {
    public static final int ENCODED_LENGTH = 26;
    public static final long MAX_TIMESTAMP = (1L << 48) - 1;

    private static final char[] ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ".toCharArray();
    private static final byte[] DECODE = new byte[128];

    static {
        Arrays.fill(DECODE, (byte) -1);
        for (int i = 0; i < ALPHABET.length; i++) {
            DECODE[ALPHABET[i]] = (byte) i;
            DECODE[Character.toLowerCase(ALPHABET[i])] = (byte) i;
        }
    }

    private final long msb;
    private final long lsb;

    private Ulid(long msb, long lsb) {
        this.msb = msb;
        this.lsb = lsb;
    }

    /** A ULID for the current time with 80 bits of cryptographic randomness. */
    public static Ulid generate() {
        byte[] r = new byte[10];
        Rng.fill(r);
        return of(System.currentTimeMillis(), r);
    }

    /** Builds a ULID from a millisecond timestamp and exactly 10 random bytes. */
    public static Ulid of(long timestampMs, byte[] random) {
        if (timestampMs < 0 || timestampMs > MAX_TIMESTAMP) {
            throw new IllegalArgumentException("timestamp must fit in 48 bits");
        }
        if (random.length != 10) {
            throw new IllegalArgumentException("random component must be 10 bytes");
        }
        long msb = timestampMs << 16 | (random[0] & 0xFFL) << 8 | (random[1] & 0xFFL);
        long lsb = 0;
        for (int i = 2; i < 10; i++) {
            lsb = lsb << 8 | (random[i] & 0xFFL);
        }
        return new Ulid(msb, lsb);
    }

    public static Ulid fromBytes(byte[] bytes) {
        if (bytes.length != 16) {
            throw new IllegalArgumentException("ULID binary form must be 16 bytes");
        }
        long msb = 0;
        long lsb = 0;
        for (int i = 0; i < 8; i++) {
            msb = msb << 8 | (bytes[i] & 0xFFL);
            lsb = lsb << 8 | (bytes[i + 8] & 0xFFL);
        }
        return new Ulid(msb, lsb);
    }

    public static Ulid fromUuid(UUID uuid) {
        return new Ulid(uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
    }

    /** Parses a canonical, case-insensitive 26-character ULID. */
    public static Ulid parse(CharSequence s) {
        if (s.length() != ENCODED_LENGTH) {
            throw new IllegalArgumentException("ULID must be 26 characters: " + s);
        }
        long hi = 0;
        long lo = 0;
        for (int i = 0; i < ENCODED_LENGTH; i++) {
            char c = s.charAt(i);
            int v = c < 128 ? DECODE[c] : -1;
            if (v < 0) {
                throw new IllegalArgumentException("invalid ULID character in: " + s);
            }
            if (i == 0 && v > 7) {
                throw new IllegalArgumentException("ULID overflows 128 bits: " + s);
            }
            hi = hi << 5 | lo >>> 59;
            lo = lo << 5 | v;
        }
        return new Ulid(hi, lo);
    }

    public long timestamp() {
        return msb >>> 16;
    }

    /** The 80-bit random component as 10 big-endian bytes. */
    public byte[] random() {
        byte[] b = toBytes();
        return Arrays.copyOfRange(b, 6, 16);
    }

    public byte[] toBytes() {
        byte[] b = new byte[16];
        for (int i = 7; i >= 0; i--) {
            b[i] = (byte) (msb >>> (8 * (7 - i)));
            b[i + 8] = (byte) (lsb >>> (8 * (7 - i)));
        }
        return b;
    }

    public UUID toUuid() {
        return new UUID(msb, lsb);
    }

    long mostSignificantBits() {
        return msb;
    }

    long leastSignificantBits() {
        return lsb;
    }

    static Ulid ofBits(long msb, long lsb) {
        return new Ulid(msb, lsb);
    }

    @Override
    public String toString() {
        char[] out = new char[ENCODED_LENGTH];
        long hi = msb;
        long lo = lsb;
        for (int i = ENCODED_LENGTH - 1; i >= 0; i--) {
            out[i] = ALPHABET[(int) (lo & 31)];
            lo = lo >>> 5 | hi << 59;
            hi >>>= 5;
        }
        return new String(out);
    }

    @Override
    public int compareTo(Ulid o) {
        int c = Long.compareUnsigned(msb, o.msb);
        return c != 0 ? c : Long.compareUnsigned(lsb, o.lsb);
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Ulid && ((Ulid) o).msb == msb && ((Ulid) o).lsb == lsb;
    }

    @Override
    public int hashCode() {
        return Long.hashCode(msb) * 31 + Long.hashCode(lsb);
    }
}

File: python/src/idgenkit/ulid.py
"""ULID: Universally Unique Lexicographically Sortable Identifier.

Spec: https://github.com/ulid/spec

    01AN4Z07BY      79KA1307SR9X4MV3
    |----------|    |----------------|
     Timestamp          Randomness
      48 bits             80 bits
"""

from __future__ import annotations

import os
import threading
import time
import uuid
from typing import Callable

__all__ = ["ULID", "MonotonicULID", "generate", "MAX_TIMESTAMP"]

CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
MAX_TIMESTAMP = (1 << 48) - 1
_RANDOM_BITS = 80
_RANDOM_MAX = (1 << _RANDOM_BITS) - 1
_MAX_INT = (1 << 128) - 1

# ULID text is Crockford base32 over a 130-bit field (2 leading zero bits).
# Encoding looks up 10 bits (two symbols) at a time. Decoding maps Crockford
# onto the digits int() accepts in base 32, after rejecting everything else
# (int() would also allow signs, underscores and whitespace).
_CROCKFORD = CROCKFORD_ALPHABET.encode("ascii")
_VALID_CHARS = _CROCKFORD + _CROCKFORD.lower()
_TO_BASE32_DIGITS = bytes.maketrans(_VALID_CHARS, b"0123456789abcdefghijklmnopqrstuv" * 2)
_PAIRS = [a + b for a in CROCKFORD_ALPHABET for b in CROCKFORD_ALPHABET]
_MASK_40 = (1 << 40) - 1


def _encode(value: int) -> str:
    p = _PAIRS
    t = value >> _RANDOM_BITS
    a = value >> 40 & _MASK_40
    b = value & _MASK_40
    return (
        f"{p[t >> 40]}{p[t >> 30 & 1023]}{p[t >> 20 & 1023]}{p[t >> 10 & 1023]}{p[t & 1023]}"
        f"{p[a >> 30]}{p[a >> 20 & 1023]}{p[a >> 10 & 1023]}{p[a & 1023]}"
        f"{p[b >> 30]}{p[b >> 20 & 1023]}{p[b >> 10 & 1023]}{p[b & 1023]}"
    )


def _decode_base32(text: str) -> int:
    """Crockford base32 (either case) to int; ValueError for any other character."""
    raw = text.encode("ascii", "replace")
    if not raw or raw.translate(None, _VALID_CHARS):
        raise ValueError(f"invalid Crockford base32 character in {text!r}")
    return int(raw.translate(_TO_BASE32_DIGITS), 32)


def _decode(text: str) -> int:
    if len(text) != 26:
        raise ValueError(f"ULID must be 26 characters, got {len(text)}")
    value = _decode_base32(text)
    if value > _MAX_INT:
        raise ValueError(f"ULID overflows 128 bits: {text!r}")
    return value


def _now_ms() -> int:
    return time.time_ns() // 1_000_000


class ULID:
    """An immutable 128-bit ULID. Ordering matches lexicographic string order."""

    __slots__ = ("_value",)

    def __init__(self, value: int) -> None:
        if not 0 <= value <= _MAX_INT:
            raise ValueError("ULID value must fit in 128 bits")
        self._value = value

    @classmethod
    def generate(cls) -> ULID:
        """Create a ULID from the current time and 80 bits of OS randomness."""
        return cls((_now_ms() << _RANDOM_BITS) | int.from_bytes(os.urandom(10), "big"))

    @classmethod
    def from_parts(cls, timestamp_ms: int, randomness: int) -> ULID:
        if not 0 <= timestamp_ms <= MAX_TIMESTAMP:
            raise ValueError("timestamp must fit in 48 bits")
        if not 0 <= randomness <= _RANDOM_MAX:
            raise ValueError("randomness must fit in 80 bits")
        return cls((timestamp_ms << _RANDOM_BITS) | randomness)

    @classmethod
    def parse(cls, text: str) -> ULID:
        return cls(_decode(text))

    @classmethod
    def from_bytes(cls, data: bytes) -> ULID:
        if len(data) != 16:
            raise ValueError("ULID binary form must be 16 bytes")
        return cls(int.from_bytes(data, "big"))

    @classmethod
    def from_uuid(cls, value: uuid.UUID) -> ULID:
        return cls(value.int)

    @property
    def timestamp_ms(self) -> int:
        return self._value >> _RANDOM_BITS

    @property
    def randomness(self) -> int:
        return self._value & _RANDOM_MAX

    def to_bytes(self) -> bytes:
        return self._value.to_bytes(16, "big")

    def to_uuid(self) -> uuid.UUID:
        return uuid.UUID(int=self._value)

    def __int__(self) -> int:
        return self._value

    def __str__(self) -> str:
        return _encode(self._value)

    def __repr__(self) -> str:
        return f"ULID({_encode(self._value)!r})"

    def __hash__(self) -> int:
        return hash(self._value)

    def __eq__(self, other: object) -> bool:
        return isinstance(other, ULID) and self._value == other._value

    def __lt__(self, other: ULID) -> bool:
        return self._value < other._value

    def __le__(self, other: ULID) -> bool:
        return self._value <= other._value

    def __gt__(self, other: ULID) -> bool:
        return self._value > other._value

    def __ge__(self, other: ULID) -> bool:
        return self._value >= other._value


class MonotonicULID:
    """Thread-safe generator whose ULIDs strictly increase.

    Within the same millisecond (or if the clock moves backwards) the previous
    randomness is incremented by one. Raises OverflowError if the 80-bit random
    component is exhausted within a single millisecond.
    """

    __slots__ = ("_lock", "_last_ms", "_last_rand")

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._last_ms = -1
        self._last_rand = 0

    def next(self) -> ULID:
        return self._next_at(_now_ms(), os.urandom)

    def _next_at(self, now: int, random: Callable[[int], bytes]) -> ULID:
        with self._lock:
            if now <= self._last_ms:
                if self._last_rand == _RANDOM_MAX:
                    raise OverflowError("ULID random component overflow within one millisecond")
                self._last_rand += 1
                now = self._last_ms
            else:
                if now > MAX_TIMESTAMP:
                    raise ValueError("timestamp must fit in 48 bits")
                self._last_ms = now
                self._last_rand = int.from_bytes(random(10), "big")
            return ULID((now << _RANDOM_BITS) | self._last_rand)


def generate() -> str:
    """Return a new random ULID as its canonical 26-character string."""
    return _encode((_now_ms() << _RANDOM_BITS) | int.from_bytes(os.urandom(10), "big"))
